### PR TITLE
Making env-runner an explicit iterator and Local Iterator regenerable 

### PR DESCRIPTION
**Pull request by Edilmo**
## Why are these changes needed?

Currently the _env_runner function is not able to handle a StopIteration properly due to be a generator created with a generator function. Here we convert env-runner in a explicit Iterator class.

Also, the LocalIterator is not able to reiterate itself after an stop generation, this is also fixed.

Some tests are being added to validate these fixes.

## Related issue number

None

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/tests/test_execution.py [hunk 1]
```diff
@@ -63,12 +63,19 @@ def step(self, action):
 class MockBaseEnv(_VectorEnvToBaseEnv):
     def __init__(self, vector_env, slow):
         super().__init__(vector_env)
-        self.slow = slow
+        import logging
+        self.logger = logging.getLogger()
```
Comment (AdityaGudimella) [inline rllib/tests/test_execution.py:67]: What is this logger being used for?

File: rllib/tests/test_execution.py [hunk 2]
```diff
@@ -204,6 +211,38 @@ def test_straggler_rollouts(ray_start_regular_shared):
...
+    a = ParallelRollouts(workers, mode="async")
+    batch = None
+    tries = 0
+    while not batch:
+        tries += 1
+        try:
+            batch = next(a)
+        except StopIteration:
+            continue
+    assert batch.count == 100
```
Comment (Random-Word) [inline rllib/tests/test_execution.py:239]: Comments explaining these test cases would help. I think I can intuit what's being established here, but it's not obvious.

**Review by AdityaGudimella — COMMENTED** [review]
LGTM!

**Review by Random-Word — APPROVED** [review]

File: rllib/tests/test_execution.py (hunk 1 above)
Comment (Edilmo) [reply rllib/tests/test_execution.py]: Nothing, it was for debugging and I forgot to delete it. Thanks. Removed

File: python/ray/util/iter.py [hunk 3]
```diff
@@ -1059,6 +1112,7 @@ def build_union(timeout=None):
                 yield_counts = [0] * len(active)
                 removed_iter_indices = []
                 for i, (weight, it) in enumerate(list(active)):
+                    expected_yield_counts = weight if weight != "*" else MAX_PULL
```
Comment (RuofanKong) [inline python/ray/util/iter.py:1115]: I don't quite follow why we move it from the case on `StopIteration`. I can sync up with you offline.

File: python/ray/util/iter.py [hunk 4]
```diff
@@ -1073,14 +1127,15 @@ def build_union(timeout=None):
                                 break
                             else:
                                 yield_counts[i] += 1
+                                if yield_counts[i] >= expected_yield_counts:
```
Comment (RuofanKong) [inline python/ray/util/iter.py:1130]: same as above comment.

File: python/ray/util/iter.py [hunk 5]
```diff
@@ -737,7 +777,13 @@ def __iter__(self):
 
     def __next__(self):
         self._build_once()
-        return next(self.built_iterator)
+        try:
+            return next(self.built_iterator)
+        except StopIteration:
+            # Force the regeneration of the base iterator
+            if self.is_infinite_sequence:
```
Comment (RuofanKong) [inline python/ray/util/iter.py:784]: I may follow up with you on the intention behind doing it.

**Review by RuofanKong — COMMENTED** [review]
@Edilmo I have a few questions especially on the intention behind the change.

**Pull request merged**